taranis/analyze_schema.py

```diff
@@ -43,70 +60,108 @@ def __init__(
+                            a_quality[record.id]["start_codon_alt"] = "alternative"
+                        except Bio.Data.CodonTable.TranslationError as e_2:
+                            if "stop" in str(e_2):
+                                a_quality[record.id]["reason"] = str(e_2).replace(
+                                    "'", ""
+                                )
+                            else:
+                                a_quality[record.id]["reason"] = str(e).replace("'", "")
+                            a_quality[record.id]["quality"] = "Bad quality"
                     else:
-                        record_sequence = str(record.seq)
-                    a_quality[record.id]["order"] = sequence_order
-                    if record_sequence[0:3] not in taranis.utils.START_CODON_FORWARD:
                         a_quality[record.id]["quality"] = "Bad quality"
-                        a_quality[record.id]["reason"] = "Start codon not found"
-                    elif record_sequence[-3:] not in taranis.utils.STOP_CODON_FORWARD:
-                        a_quality[record.id]["quality"] = "Bad quality"
-                        a_quality[record.id]["reason"] = "Stop codon not found"
-                    
-                    elif taranis.utils.find_multiple_stop_codons(record_sequence):
-                        a_quality[record.id]["quality"] = "Bad quality"
-                        a_quality[record.id]["reason"] = "Multiple stop codons found"
+                        a_quality[record.id]["reason"] = str(e).replace("'", "")
+
                 if (
                     self.remove_no_cds
                     and a_quality[record.id]["quality"] == "Bad quality"
                 ):
                     bad_quality_record.append(record.id)
 
-        if self.remove_duplicated:
-            # get the unique sequences and compare the length with all sequences
-            unique_seq = list(set(list(allele_seq.values())))
-            if len(unique_seq) < len(allele_seq):
-                tmp_dict = {}
-                for rec_id, seq_value in allele_seq.items():
-                    if seq_value not in tmp_dict:
-                        tmp_dict[seq_value] = 0
-                    else:
-                        bad_quality_record.append(rec_id)
-                        a_quality[rec_id]["quality"] ="Bad quality"
-                        a_quality[rec_id]["reason"] ="Duplicate allele"
-        if self.remove_subset:
-            unique_seq = list(set(list(allele_seq.values())))
+        # check if there are duplicated alleles
+        # get the unique sequences and compare the length with all sequences
+        unique_seq = list(set(list(allele_seq.values())))
+        if len(unique_seq) < len(allele_seq):
+            tmp_dict = {}
             for rec_id, seq_value in allele_seq.items():
-                unique_seq.remove(seq_value)
-                if seq_value in unique_seq:
+                if seq_value not in tmp_dict:
```

Comment:
I feell that maybe this can be done without a tmp_dict?
Something like this?
```
 value_to_keys = defaultdict(list)

    # Iterate over items and group keys by value
    for key, value in my_dict.items():
        value_to_keys[value].append(key)

    # Filter out values with only one key (no duplicates)
    duplicates = {value: keys for value, keys in value_to_keys.items() if len(keys) > 1}
``` 

And the use the duplicates to create the a_quality

Reply:
yes, code is nicer for getting only which are duplicated, however, I need to create another loop for getting the allele_id based on the sequence value to update the a_quality dictionary for setting the bad quality reason, which requires more CPU time. Because it should be like this:
```
value_to_keys = defaultdict(list)

# Iterate over items and group keys by value
for key, value in my_dict.items():
    value_to_keys[value].append(key)

# Filter out values with only one key (no duplicates)
duplicates = {value: keys for value, keys in value_to_keys.items() if len(keys) > 1}

# Additional code
for seq_id, seq_value in allele_seq.items():
      if seq_value in duplicates:
          for id in duplicates[seq_value]:
              a_quality[id]["quality"] = "Bad quality"
              a_quality[id]["reason"] = "Duplicate allele"
```
Please correct me if I misunderstood.

Reply:
mm don't understand, if you keep the seq_id in the duplicates list, you don't need to iterate over that, right? 
Maybe you need to twist a little this in order to keep the seq_ids instead of the seq_values:
```
duplicates = {value: keys for value, keys in value_to_keys.items() if len(keys) > 1}
```

Reply:
I was thinking on this code

```
            value_to_keys = defaultdict(list)
            for rec_id, seq_value in allele_seq.items():
                value_to_keys[seq_value].append(rec_id)
                if len(value_to_keys[seq_value]) > 1:
                    a_quality[rec_id]["quality"] = "Bad quality"
                    a_quality[rec_id]["reason"] = "Duplicate allele"
                    if self.remove_duplicated:
                        bad_quality_record.append(rec_id)
```

Reply:
Updated this part of code having only 1 loop and without any temporary dictionary